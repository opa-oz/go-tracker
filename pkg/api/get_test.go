package api_test

import (
	"github.com/gkampitakis/go-snaps/snaps"
	"net/http"
	"net/http/httptest"
	"testing"
	"tracker/pkg/api"
	"tracker/pkg/middlewares"

	"github.com/gin-gonic/gin"
	"github.com/go-playground/assert/v2"
	"github.com/go-redis/redismock/v9"
)

func TestGetRoute(t *testing.T) {
	r := gin.Default()

	rdb, mock := redismock.NewClientMock()

	r.Use(middlewares.RedisMiddleware(rdb))
	r.GET("/get/:name", api.Get)
	r.GET("/g/:name", api.Get)

	t.Run("if keys is not set, get returns 0", func(t *testing.T) {
		mock.ExpectGet("test").RedisNil()
		mock.ExpectSet("test", 0, 0).SetVal("0")

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/get/test", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"value":0}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("if key is set, get returns value", func(t *testing.T) {
		value := "100"
		mock.ExpectGet("test").SetVal(value)

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/g/test", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"value":100}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})

	t.Run("if max is set, get returns less or equal", func(t *testing.T) {
		value := "100"
		mock.ExpectGet("test").SetVal(value)

		w := httptest.NewRecorder()
		req, _ := http.NewRequest(http.MethodGet, "/get/test?max=50", nil)
		r.ServeHTTP(w, req)

		assert.Equal(t, `{"value":50}`, w.Body.String())
		snaps.MatchSnapshot(t, w.Body.String())
		assert.Equal(t, http.StatusOK, w.Code)
	})
}
