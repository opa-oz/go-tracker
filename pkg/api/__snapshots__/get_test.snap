
[TestGetRoute/if_keys_is_not_set,_get_returns_0 - 1]
{"value":0}
---

[TestGetRoute/if_key_is_set,_get_returns_value - 1]
{"value":100}
---

[TestGetRoute/if_max_is_set,_get_returns_less_or_equal - 1]
{"value":50}
---
